File: UnitySample/Assets/Samples/AppleIdLogin/AppleIdLoginTest.cs
// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: MIT-0

using System.Collections;
using System.Collections.Generic;
using UnityEngine;
using UnityEngine.Networking;
using UnityEngine.UI;

public class AppleIdLoginTest : MonoBehaviour
{
    public string loginEndpointUrl;
    public Text logOutput;

    // Start is called before the first frame update
    void Start()
    {
        // Set the login endpoint
        Debug.Log("Setting login endpoint");
        AWSGameSDKClient.Instance.Init(loginEndpointUrl, this.OnLoginOrRefreshError);

        // Get a new guest user first to test account linking
        Debug.Log("Requesting new guest identity first");
        this.logOutput.text += "Requesting new identity\n";
        AWSGameSDKClient.Instance.LoginAsNewGuestUser(this.OnGuestLoginResponse);
    }

    // Triggered by the SDK if there's any login error or error refreshing access token
    void OnLoginOrRefreshError(string error)
    {
        Debug.LogError("Login or refresh error: " + error);
        this.logOutput.text += "Login or refresh error: " + error + "\n";

        // NOTE: You would here trigger a new log in or other remediation
    }

    // Update is called once per frame
    void Update()
    {
        
    }

    void OnGuestLoginResponse(LoginRequestData userInfo)
    {
        Debug.Log("New guest Login response: UserID: " + userInfo.user_id + "GuestSecret: " + userInfo.guest_secret + "Token: " + userInfo.auth_token);
        this.logOutput.text += "Login response: \nUserID: " + userInfo.user_id + " \nGuestSecret: " + userInfo.guest_secret + " \n";
        
        // Now let's test linking Apple ID
        // NOTE: You're expected to input a valid Apple ID token here, see the documentation for Unity Apple Sign in for details on how to do this: https://docs.unity.com/authentication/en/manual/set-up-apple-signin
        AWSGameSDKClient.Instance.LinkAppleIdToCurrentUser("eyYourTokenHere", this.OnLinkAppleIdResponse);
    }

    void OnLinkAppleIdResponse(LoginRequestData userInfo)
    {
        Debug.Log("Apple ID linking response: UserID: " + userInfo.user_id + "AppleId: " + userInfo.apple_id + "Token: " + userInfo.auth_token);
        this.logOutput.text += "Apple ID linking response: \nUserID: " + userInfo.user_id + "\nAppleId: " + userInfo.apple_id + " \n";

        // Now let's test logging in with this existing Apple ID
        // NOTE: As above, you're expected to have a valid Apple ID token here
        AWSGameSDKClient.Instance.LoginWithAppleIdToken("eyYourTokenHere", this.OnLoginWithAppleIdResponse);
    }

    void OnLoginWithAppleIdResponse(LoginRequestData userInfo)
    {
        Debug.Log("Apple ID login response: UserID: " + userInfo.user_id + "AppleId: " + userInfo.apple_id + "Token: " + userInfo.auth_token);
        this.logOutput.text += "Apple ID login response: \nUserID: " + userInfo.user_id + "\nAppleId: " + userInfo.apple_id + " \nToken: " + userInfo.auth_token + "\n";
    }
}

File: UnrealSample/Source/UnrealSample/Samples/AppleIdLoginTest/AppleIdLoginTest.cpp
// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: MIT-0

#include "AppleIdLoginTest.h"
#include "../../AWSGameSDK/AWSGameSDK.h"
#include "Engine/World.h"
#include "Kismet/GameplayStatics.h"

// Sets default values for this component's properties
UAppleIdLoginTest::UAppleIdLoginTest()
{
	// Set this component to be initialized when the game starts, and to be ticked every frame.  You can turn these features
	// off to improve performance if you don't need them.
	PrimaryComponentTick.bCanEverTick = true;

	// ...
}


// Called when the game starts
void UAppleIdLoginTest::BeginPlay()
{
	Super::BeginPlay();

	// Get the subsystems
    UGameInstance* GameInstance = Cast<UGameInstance>(UGameplayStatics::GetGameInstance(GetWorld()));
    UAWSGameSDK* AWSGameSDK =  GameInstance->GetSubsystem<UAWSGameSDK>();

    // Init with the login endpoint defined in the Editor and a callback to handle errors for logging in and refresh
	AWSGameSDK->Init(this->m_loginEndpoint);
	AWSGameSDK->OnLoginFailure.AddUObject(this, &UAppleIdLoginTest::OnLoginOrRefreshErrorCallback);

	// Define the OnLoginResult callback
	UAWSGameSDK::FLoginComplete loginCallback;
	loginCallback.BindUObject(this, &UAppleIdLoginTest::OnGuestLoginResultCallback);

	// Login as a new guest user first
    AWSGameSDK->LoginAsNewGuestUser(loginCallback);
	
}

// Called when there is an error with login or token refresh. You will need to handle logging in again here
void UAppleIdLoginTest::OnLoginOrRefreshErrorCallback(const FString& errorMessage){
    UE_LOG(LogTemp, Display, TEXT("Received login error: %s \n"), *errorMessage);
    if(GEngine)
        GEngine->AddOnScreenDebugMessage(-1, 30.0f, FColor::Red, FString::Printf(TEXT("Received login error: \n %s \n"), *errorMessage), false, FVector2D(1.5f,1.5f));

    // NOTE:  You will need to handle logging in again here
}

// Called when guest login is done
void UAppleIdLoginTest::OnGuestLoginResultCallback(const UserInfo& userInfo){
    UE_LOG(LogTemp, Display, TEXT("Received guest login response: %s \n"), *userInfo.ToString());
    if(GEngine)
            GEngine->AddOnScreenDebugMessage(-1, 30.0f, FColor::Black, FString::Printf(TEXT("Received login response: \n %s \n"), *userInfo.ToString()), false, FVector2D(1.5f,1.5f));

    // Test linking apple id to the existing guest account
	UAWSGameSDK::FLoginComplete onLinkAppleIdCallback;
	onLinkAppleIdCallback.BindUObject(this, &UAppleIdLoginTest::OnLinkAppleIdResultCallback);
	UGameInstance* GameInstance = Cast<UGameInstance>(UGameplayStatics::GetGameInstance(GetWorld()));
    UAWSGameSDK* AWSGameSDK = GameInstance->GetSubsystem<UAWSGameSDK>();

    // NOTE: You need to use an appropriate Apple SDK plugin to sign in with Apple and get the Apple ID token for the user and pass it here
    AWSGameSDK->LinkAppleIdToCurrentUser("eyYourToken", onLinkAppleIdCallback);
}

void UAppleIdLoginTest::OnLinkAppleIdResultCallback(const UserInfo& userInfo){

    UE_LOG(LogTemp, Display, TEXT("Received apple ID linking response: %s \n"), *userInfo.ToString());
    if(GEngine)
            GEngine->AddOnScreenDebugMessage(-1, 30.0f, FColor::Black, FString::Printf(TEXT("Received apple ID linking response: \n %s \n"), *userInfo.ToString()), false, FVector2D(1.5f,1.5f));

    // Test logging in with existing apple_id
	UAWSGameSDK::FLoginComplete onAppleIdLoginCallback;
	onAppleIdLoginCallback.BindUObject(this, &UAppleIdLoginTest::OnLoginWithAppleId);
	UGameInstance* GameInstance = Cast<UGameInstance>(UGameplayStatics::GetGameInstance(GetWorld()));
    UAWSGameSDK* AWSGameSDK = GameInstance->GetSubsystem<UAWSGameSDK>();

    // NOTE: You need to use an appropriate Apple SDK plugin to sign in with Apple and get the Apple ID token for the user and pass it here
    AWSGameSDK->LoginWithAppleIdToken("eyYourToken", onAppleIdLoginCallback);
}

void UAppleIdLoginTest::OnLoginWithAppleId(const UserInfo& userInfo){

    UE_LOG(LogTemp, Display, TEXT("Received apple ID login response: %s \n"), *userInfo.ToString());
    if(GEngine)
        GEngine->AddOnScreenDebugMessage(-1, 30.0f, FColor::Black, FString::Printf(TEXT("Received apple ID login response: \n %s \n"), *userInfo.ToString()), false, FVector2D(1.5f,1.5f));
}

// Called every frame
void UAppleIdLoginTest::TickComponent(float DeltaTime, ELevelTick TickType, FActorComponentTickFunction* ThisTickFunction)
{
	Super::TickComponent(DeltaTime, TickType, ThisTickFunction);

	// ...
}

